File: kotlin/workflow-core/src/main/java/com/squareup/workflow/WorkflowAction.kt
/*
 * Copyright 2019 Square Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.workflow

import com.squareup.workflow.WorkflowAction.Companion.toString
import com.squareup.workflow.WorkflowAction.Updater

/**
 * An atomic operation that updates the state of a [Workflow], and also optionally emits an output.
 */
interface WorkflowAction<StateT, out OutputT : Any> {
  @Deprecated("Use Updater")
  class Mutator<S>(var state: S)

  /**
   * The context for calls to [WorkflowAction.apply]. Allows the action to set the
   * [nextState], and to emit the [setOutput].
   *
   * @param nextState the state that the workflow should move to. Default is the current state.
   */
  class Updater<S, in O : Any>(var nextState: S) {
    internal var output: @UnsafeVariance O? = null

    /**
     * Sets the value the workflow will emit as output when this action is applied.
     * If this method is not called, there will be no output.
     */
    fun setOutput(output: O) {
      this.output = output
    }
  }

  @Suppress("DEPRECATION")
  @Deprecated("Implement Updater.apply")
  fun Mutator<StateT>.apply(): OutputT? {
    throw UnsupportedOperationException()
  }

  @Suppress("DEPRECATION")
  fun Updater<StateT, OutputT>.apply() {
    val mutator = Mutator(nextState)
    mutator.apply()
        ?.let { setOutput(it) }
    nextState = mutator.state
  }

  companion object {
    /**
     * Returns a [WorkflowAction] that does nothing: no output will be emitted, and
     * the state will not change.
     *
     * Use this to, for example, ignore the output of a child workflow or worker.
     */
    @Suppress("UNCHECKED_CAST")
    fun <StateT, OutputT : Any> noAction(): WorkflowAction<StateT, OutputT> =
      NO_ACTION as WorkflowAction<StateT, OutputT>

    /**
     * Convenience function that returns a [WorkflowAction] that will just set the state to [newState]
     * (without considering the current state) and optionally emit an output.
     */
    @Deprecated(
        message = "Use action",
        replaceWith = ReplaceWith(
            expression = "action { nextState = newState }",
            imports = arrayOf("com.squareup.workflow.action")
        )
    )
    fun <StateT, OutputT : Any> enterState(
      newState: StateT,
      emittingOutput: OutputT? = null
    ): WorkflowAction<StateT, OutputT> =
      action({ "enterState($newState, $emittingOutput)" }) {
        nextState = newState
        emittingOutput?.let { setOutput(it) }
      }

    /**
     * Convenience function that returns a [WorkflowAction] that will just set the state to [newState]
     * (without considering the current state) and optionally emit an output.
     */
    @Deprecated(
        message = "Use action",
        replaceWith = ReplaceWith(
            expression = "action { nextState = newState }",
            imports = arrayOf("com.squareup.workflow.action")
        )
    )
    fun <StateT, OutputT : Any> enterState(
      name: String,
      newState: StateT,
      emittingOutput: OutputT? = null
    ): WorkflowAction<StateT, OutputT> =
      action({ "enterState($name, $newState, $emittingOutput)" }) {
        nextState = newState
        emittingOutput?.let { setOutput(it) }
      }

    /**
     * Convenience function to implement [WorkflowAction] without returning the output.
     */
    @Deprecated(
        message = "Use action",
        replaceWith = ReplaceWith(
            expression = "action(name) { nextState = state }",
            imports = arrayOf("com.squareup.workflow.action")
        )
    )
    fun <StateT, OutputT : Any> modifyState(
      name: () -> String,
      emittingOutput: OutputT? = null,
      modify: (StateT) -> StateT
    ): WorkflowAction<StateT, OutputT> =
      action({ "modifyState(${name()}, $emittingOutput)" }) {
        nextState = modify(nextState)
        emittingOutput?.let { setOutput(it) }
      }

    /**
     * Convenience function to implement [WorkflowAction] without changing the state.
     */
    @Deprecated(
        message = "Use action",
        replaceWith = ReplaceWith(
            expression = "action { setOutput(output) }",
            imports = arrayOf("com.squareup.workflow.action")
        )
    )
    fun <StateT, OutputT : Any> emitOutput(output: OutputT): WorkflowAction<StateT, OutputT> =
      action({ "emitOutput($output)" }) { setOutput(output) }

    /**
     * Convenience function to implement [WorkflowAction] without changing the state.
     */
    @Deprecated(
        message = "Use action",
        replaceWith = ReplaceWith(
            expression = "action { setOutput(output) }",
            imports = arrayOf("com.squareup.workflow.action")
        )
    )
    fun <StateT, OutputT : Any> emitOutput(
      name: String,
      output: OutputT
    ): WorkflowAction<StateT, OutputT> =
      action({ "emitOutput($name, $output)" }) { setOutput(output) }

    private val NO_ACTION = action<Any, Any>({ "noAction" }) { }
  }
}

/**
 * Creates a [WorkflowAction] from the [apply] lambda.
 * The returned object will include the string returned from [name] in its [toString].
 *
 * If defining actions within a [StatefulWorkflow], use the [StatefulWorkflow.workflowAction]
 * extension instead, to do this without being forced to repeat its parameter types.
 *
 * @param name A string describing the update for debugging.
 * @param apply Function that defines the workflow update.
 *
 * @see StatelessWorkflow.action
 * @see StatefulWorkflow.action
 */
inline fun <StateT, OutputT : Any> action(
  name: String = "",
  crossinline apply: Updater<StateT, OutputT>.() -> Unit
) = action({ name }, apply)

/**
 * Creates a [WorkflowAction] from the [apply] lambda.
 * The returned object will include the string returned from [name] in its [toString].
 *
 * If defining actions within a [StatefulWorkflow], use the [StatefulWorkflow.workflowAction]
 * extension instead, to do this without being forced to repeat its parameter types.
 *
 * @param name Function that returns a string describing the update for debugging.
 * @param apply Function that defines the workflow update.
 *
 * @see StatelessWorkflow.action
 * @see StatefulWorkflow.action
 */
inline fun <StateT, OutputT : Any> action(
  crossinline name: () -> String,
  crossinline apply: Updater<StateT, OutputT>.() -> Unit
): WorkflowAction<StateT, OutputT> = object : WorkflowAction<StateT, OutputT> {
  override fun Updater<StateT, OutputT>.apply() = apply.invoke(this)
  override fun toString(): String = "WorkflowAction(${name()})@${hashCode()}"
}

fun <StateT, OutputT : Any> WorkflowAction<StateT, OutputT>.applyTo(
  state: StateT
): Pair<StateT, OutputT?> {
  val updater = Updater<StateT, OutputT>(state)
  updater.apply()
  return Pair(updater.nextState, updater.output)
}

File: swift/Workflow/Sources/WorkflowAction.swift
/*
 * Copyright 2020 Square Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/// Conforming types represent an action that advances a workflow. When applied, an action emits the next
/// state and / or output for the workflow.
public protocol WorkflowAction {
    /// The type of workflow that this action can be applied to.
    associatedtype WorkflowType: Workflow

    /// Applies this action to a given state of the workflow, optionally returning an output event.
    ///
    /// - Parameter state: The current state of the workflow. The state is passed as an `inout` param, allowing actions
    ///                    to modify state during application.
    ///
    /// - Returns: An optional output event for the workflow. If an output event is returned, it will be passed up
    ///            the workflow hierarchy to this workflow's parent.
    func apply(toState state: inout WorkflowType.State) -> WorkflowType.Output?
}

/// A type-erased workflow action.
///
/// The `AnyWorkflowAction` type forwards `apply` to an underlying workflow action, hiding its specific underlying type,
/// or to a closure that implements the `apply` logic.
public struct AnyWorkflowAction<WorkflowType: Workflow>: WorkflowAction {
    private let _apply: (inout WorkflowType.State) -> WorkflowType.Output?

    /// Creates a type-erased workflow action that wraps the given instance.
    ///
    /// - Parameter base: A workflow action to wrap.
    public init<E>(_ base: E) where E: WorkflowAction, E.WorkflowType == WorkflowType {
        if let anyEvent = base as? AnyWorkflowAction<WorkflowType> {
            self = anyEvent
            return
        }
        self._apply = { return base.apply(toState: &$0) }
    }

    /// Creates a type-erased workflow action with the given `apply` implementation.
    ///
    /// - Parameter apply: the apply function for the resulting action.
    public init(_ apply: @escaping (inout WorkflowType.State) -> WorkflowType.Output?) {
        self._apply = apply
    }

    public func apply(toState state: inout WorkflowType.State) -> WorkflowType.Output? {
        return _apply(&state)
    }
}

extension AnyWorkflowAction {
    /// Creates a type-erased workflow action that simply sends the given output event.
    ///
    /// - Parameter output: The output event to send when this action is applied.
    public init(sendingOutput output: WorkflowType.Output) {
        self = AnyWorkflowAction { state in
            output
        }
    }

    /// Creates a type-erased workflow action that does nothing (it leaves state unchanged and does not emit an output
    /// event).
    public static var noAction: AnyWorkflowAction<WorkflowType> {
        return AnyWorkflowAction { state in
            nil
        }
    }
}
